### PR TITLE
CORE-7646 Added support in the virtual node writer service for creating nodes async

### PR DESCRIPTION
**Pull request by owenstanford**
Added support in the virtual node writer service for creating nodes async

Corresponding `corda-api` chnage: https://github.com/corda/corda-api/pull/891

### CODE REVIEW

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/v1/VirtualNodeRestResourceImpl.kt [hunk 1]
```diff
@@ -410,13 +394,10 @@ internal class VirtualNodeRestResourceImpl @Activate constructor(
...
         CpiIdentifier(name, version, signerSummaryHash.toString())
 
-    /** Validates the [x500Name]. */
-    private fun validateX500Name(x500Name: String) = try {
-        MemberX500Name.parse(x500Name)
-    } catch (e: Exception) {
-        logger.warn("Configuration Management  X500 name \"$x500Name\" could not be parsed. Cause: ${e.message}")
-        val message = "X500 name \"$x500Name\" could not be parsed. Cause: ${e.message}"
-        throw InvalidInputDataException(message)
+    private fun isRunningGuard() {
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/v1/VirtualNodeRestResourceImpl.kt]: We actually shouldn't need this check at all because the method invoker set up on the Javalin route handler already checks this Lifecycle component is running before invoking the method:

https://github.com/corda/corda-runtime-os/blob/0802fa959453c2341c706a8abc6648433e471d2c/libs/rest/rest-server-impl/src/main/kotlin/net/corda/httprpc/server/impl/apigen/processing/MethodInvoker.kt#L36-L41

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt [hunk 2]
```diff
@@ -2,19 +2,105 @@ package net.corda.virtualnode.rpcops.impl.validation.impl
...
+import java.util.UUID
 
 internal class VirtualNodeValidationServiceImpl(
     private val virtualNodeInfoReadService: VirtualNodeInfoReadService,
     private val cpiInfoReadService: CpiInfoReadService
-): VirtualNodeValidationService {
+) : VirtualNodeValidationService {
+
+    override fun validateVirtualNodeDoesNotExist(holdingIdentity: HoldingIdentity) {
+        if(virtualNodeInfoReadService.getByHoldingIdentityShortHash(holdingIdentity.shortHash)!=null){
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt]: Probably could do with reformatting this class

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt [hunk 3]
```diff
@@ -2,19 +2,105 @@ package net.corda.virtualnode.rpcops.impl.validation.impl
...
+            throw ResourceAlreadyExistsException(
+                "A virtual node for '$holdingIdentity' already exists."
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt]: What does toString() on this holdingIdentity return?

Maybe should use the other constructor:

```
throw ResourceAlreadyExistsException("Virtual Node", holdingIdentity.shortHash.toString())
```

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt (hunk 2 above)
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt]: Minor:
```suggestion
        if (virtualNodeInfoReadService.getByHoldingIdentityShortHash(holdingIdentity.shortHash) != null) {
```

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt [hunk 4]
```diff
@@ -2,19 +2,105 @@ package net.corda.virtualnode.rpcops.impl.validation.impl
...
+            )
+        }
+    }
+
+    override fun validateAndGetGroupId(request: VirtualNodeRequest):String {
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt]: I think this `VirtualNodeRequest` really ought to be called VirtualNodeCreationRequest otherwise it's a bit ambiguous, though I appreciate this could be a significant increase in the PR's footprint, so I'm happy if you don't make that change.

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/RequestFactory.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import net.corda.libs.virtualnode.endpoints.v1.types.VirtualNodeRequest
+import net.corda.virtualnode.HoldingIdentity
+
+interface RequestFactory {
+    fun createHoldingIdentity(groupId: String, request: VirtualNodeRequest): HoldingIdentity
+
+    fun createVirtualNodeRequest(
+        holdingIdentity: HoldingIdentity,
+
+        request: VirtualNodeRequest
```
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/RequestFactory.kt]: Minor:
```suggestion
        holdingIdentity: HoldingIdentity,
        request: VirtualNodeRequest
```

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt [hunk 6]
```diff
@@ -2,19 +2,105 @@ package net.corda.virtualnode.rpcops.impl.validation.impl
...
+        } catch (e: GroupPolicyParseException) {
+            throw InternalServerException("Could not find group ID in CPI policy data '${groupPolicyJson}'")
+        }
+
+        // generate a group ID when creating a virtual node for an MGM default group.
+        return if (groupId == GroupPolicyConstants.PolicyValues.Root.MGM_DEFAULT_GROUP_ID) {
+            UUID.randomUUID().toString()
+        } else {
+            groupId
+        }
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt:92]: 👍 

So this avoids having to go to db-worker to obtain the groupId which is then used to calculate the holdingIdentityShortHash.

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/RequestFactory.kt [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    ): VirtualNodeAsynchronousRequest
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/RequestFactory.kt:14]: Formatting / empty line here too

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/CpiEntityRepository.kt [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package net.corda.virtualnode.write.db.impl.writer
+
+import javax.persistence.EntityManager
+
+internal interface CpiEntityRepository {
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/CpiEntityRepository.kt:5]: Just making aware of potential conflicts with https://r3-cev.atlassian.net/browse/CORE-8744

But yes introducing this interface may help refactoring in CORE-8744

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt [hunk 9]
```diff
@@ -34,11 +35,9 @@ internal class VirtualNodeWriterFactory(
     private val dbConnectionManager: DbConnectionManager,
     private val virtualNodeDbAdmin: VirtualNodesDbAdmin,
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt]: virtualNodeDbAdmin seems not to be used? But leave schemaMigrator as I have a PR to fix flaky test which changes this code to use this injected schemaMigrator:

https://github.com/corda/corda-runtime-os/pull/3235

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt [hunk 10]
```diff
@@ -69,13 +68,19 @@ internal class VirtualNodeWriterFactory(
             VirtualNodeEntityRepository(this.dbConnectionManager.getClusterEntityManagerFactory())
         val migrationUtility = MigrationUtilityImpl(dbConnectionManager, LiquibaseSchemaMigratorImpl())
 
-        val virtualNodeUpgradeHandler = VirtualNodeUpgradeOperationHandler(
-            this.dbConnectionManager.getClusterEntityManagerFactory(),
-            oldVirtualNodeEntityRepository,
-            virtualNodeInfoPublisher,
-            migrationUtility
+        val handlerMap = mutableMapOf<Class<*>, VirtualNodeAsyncOperationHandler<*>>(
+            VirtualNodeUpgradeRequest::class.java to VirtualNodeUpgradeOperationHandler(
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt:94]: I think you forgot to add your CreateVirtualNodeOperationHandler?

That's one reason why injecting each handler to `VirtualNodeAsyncOperationProcessor` gives us some compile-time safety. But I'm Ok with this refactoring since this pattern is in line with what we do in the flow pipeline handlers, etc.

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/services/CreateVirtualNodeServiceImpl.kt [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+                return "If Uniqueness DDL connection is provided, Uniqueness DML connection needs to be provided as well."
+            }
+        }
+        return null
+    }
+
+    override fun ensureHoldingIdentityIsUnique(request: VirtualNodeCreateRequest) {
+        val emf = dbConnectionManager.getClusterEntityManagerFactory()
+        val holdingIdShortHash = request.holdingId.toCorda().shortHash
+        emf.use { em ->
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/services/CreateVirtualNodeServiceImpl.kt:76]: Should we be performing these reads in a transaction?

`emf.transaction { ...`

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/services/CreateVirtualNodeServiceImpl.kt [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        return cpiEntityRepository.getCpiMetadataByChecksum(cpiFileChecksum)
+            ?: throw  CpiNotFoundException("CPI with file checksum ${cpiFileChecksum} was not found.")
+    }
+
+    override fun runCpiMigrations(
+        cpiMetadata: CpiMetadataLite,
+        vaultDb: VirtualNodeDb,
+        holdingIdentity: HoldingIdentity
+    ) {
+        dbConnectionManager.getClusterEntityManagerFactory().createEntityManager().transaction { em ->
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/services/CreateVirtualNodeServiceImpl.kt:102]: Not part of this change, but I created https://r3-cev.atlassian.net/browse/CORE-10935 to fix an issue where the CPI migrations fail and the x500 name cannot create virtual nodes in the same group.

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/services/CreateVirtualNodeServiceImpl.kt [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+import net.corda.virtualnode.write.db.impl.writer.VirtualNodeDbChangeLog
+import net.corda.virtualnode.write.db.impl.writer.VirtualNodeDbConnections
+import org.slf4j.LoggerFactory
+import java.util.UUID
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+import javax.persistence.EntityManager
+
+@Suppress("LongParameterList")
+internal class CreateVirtualNodeServiceImpl(
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/services/CreateVirtualNodeServiceImpl.kt:38]: Yeah I like this separation of publishing / creation of transactions / interactions with repositories into a service class. I'll refactor the upgrade handler if I get a chance. One thing I'd suggest is since "service" is a common name I wonder if there is a better name... mediator / composer / ? I'll leave that up to you to decide if you want to change it.

File: libs/virtual-node/virtual-node-endpoints/src/main/kotlin/net/corda/libs/virtualnode/endpoints/v1/VirtualNodeRestResource.kt [hunk 14]
```diff
@@ -37,7 +36,7 @@ interface VirtualNodeRestResource : RestResource {
     fun createVirtualNode(
         @ClientRequestBodyParameter(description = "Details of the virtual node to be created")
         request: VirtualNodeRequest
-    ): VirtualNodeInfo
+    ): ResponseEntity<AsyncResponse>
```
Comment (sprintengine) [inline libs/virtual-node/virtual-node-endpoints/src/main/kotlin/net/corda/libs/virtualnode/endpoints/v1/VirtualNodeRestResource.kt:40]: You'll need a change to the swagger baseline following this change.

There is a readme on best approach to do that - 

https://github.com/corda/corda-runtime-os/blob/a1c5fa4a40ec22a38e04e8d700832d58ff61fb30/processors/rest-processor/src/integrationTest/resources/readme.md#L1-L22

**Review by conal-smith — COMMENTED** [review]
I like the approaches taken, this looks good overall

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt (hunk 9 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt]: this was linked to the missing handler (comment below), it is used now.

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt (hunk 10 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt:94]: Yes this was missing (added now)!
I think there has been a recognition (by most) that we have gone a bit OTT with our overuse of OSGI for DI and it's starting to hurting us. To make the handler injectable by OSGI we would need to attribute all the internal classes we inject which would add to the OSGI proliferation. while the manual boilerplate is a pain I think it's the least bad option.

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/CpiEntityRepository.kt (hunk 8 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/CpiEntityRepository.kt:5]: yes this is a stop gap to allow the old concrete type to be injected.

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/v1/VirtualNodeRestResourceImpl.kt (hunk 1 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/v1/VirtualNodeRestResourceImpl.kt]: ah good! I didn't realize this was already there, this also means I can close the tech-debt I raised to implement it :)

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt (hunk 10 above)
Comment (sprintengine) [reply components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt:94]: Yes, we definitely have OTT use of OSGi for DI. The vnode operation handlers and `VirtualNodeAsyncOperationProcessor` are not OSGi components, and I don't want to make them OSGI components. IMO we don't need to pass in a `handlerMap` as we lose type safety. 

> But I'm Ok with this refactoring since this pattern is in line with what we do in the flow pipeline handlers, etc.

I notice the patterns used in places like `FlowEventPipelineFactoryImpl` are there to deals with OSGi components. But we don't have OSGi components here and we can just simple Java constructor dependency injection for the virtualNodeUpgradeHandler and virtualNodeCreationHandler.

What I mean is, passing in a `handlerMap` is useful when we dynamically load implementations at runtime (i.e. if we're using OSGi), but these handlers aren't OSGi, there's only two of them, and we can just inject each one into the constructor of the `VirtualNodeAsyncOperationProcessor`.

But I guess a benefit of using a handlerMap is you just need to change the VirtualNodeWriterFactory in order to introduce new handlers. I'm easy for whichever way you prefer.

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt (hunk 3 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt]: toString() returns the x500 name and group id, changed to use the alternative constructor.

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt (hunk 4 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/validation/impl/VirtualNodeValidationServiceImpl.kt]: updated

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt (hunk 10 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterFactory.kt:94]: Yes the handlerMap is a lazy handler factory/provider, If we're going with a map It's better to pass it in IMO, it makes unit testing and extending a little easier and puts the construction and mapping in one place. I'm not a fan of wide constructors or passing in list type data as constructor args, but I do see the point you're making.

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/services/CreateVirtualNodeServiceImpl.kt (hunk 11 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/services/CreateVirtualNodeServiceImpl.kt:76]: No, for this point in time check of existence it's not going to add anything

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ClusterBootstrapTest.kt [hunk 15]
```diff
@@ -35,7 +35,7 @@ class ClusterBootstrapTest {
     )
     private val client = HttpClient.newBuilder().build()
 
-    @Test
+    //@Test
```
Comment (vkolomeyko) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ClusterBootstrapTest.kt]: We probably should restore:
```suggestion
    @Test
```

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/RequestFactory.kt [hunk 16]
```diff
@@ -0,0 +1,15 @@
+package net.corda.virtualnode.rpcops.factories
+
+import net.corda.data.virtualnode.VirtualNodeAsynchronousRequest
+import net.corda.libs.virtualnode.endpoints.v1.types.CreateVirtualNodeRequest
+import net.corda.virtualnode.HoldingIdentity
+
+interface RequestFactory {
```
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/RequestFactory.kt]: Just to highlight the fact that there is no intention to use it outside of this module:
```suggestion
internal interface RequestFactory {
```

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/impl/RequestFactoryImpl.kt [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+import net.corda.data.virtualnode.VirtualNodeCreateRequest
+import net.corda.rest.security.RestContextProvider
+import net.corda.libs.virtualnode.endpoints.v1.types.CreateVirtualNodeRequest
+import net.corda.utilities.time.Clock
+import net.corda.v5.base.types.MemberX500Name
+import net.corda.virtualnode.HoldingIdentity
+import net.corda.virtualnode.rpcops.factories.RequestFactory
+import net.corda.virtualnode.toAvro
+
+class RequestFactoryImpl(
```
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/impl/RequestFactoryImpl.kt]:
```suggestion
internal class RequestFactoryImpl(
```

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ClusterBootstrapTest.kt (hunk 15 above)
Comment (owenstanford) [reply applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ClusterBootstrapTest.kt]: fixed

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/CreateVirtualNodeOperationHandler.kt [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    private val policyParser: GroupPolicyParser,
+    private val logger: Logger
+) : VirtualNodeAsyncOperationHandler<VirtualNodeCreateRequest> {
+
+    override fun handle(
+        requestTimestamp: Instant,
+        requestId: String,
+        request: VirtualNodeCreateRequest
+    ) {
+        val holdingId =request.holdingId.toCorda()
```
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/CreateVirtualNodeOperationHandler.kt]: Nit:
```suggestion
        val holdingId = request.holdingId.toCorda()
```

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/VirtualNodeUpgradeOperationHandler.kt [hunk 19]
```diff
@@ -60,7 +60,7 @@ internal class VirtualNodeUpgradeOperationHandler(
             handleMigrationsFailed(request, requestId, requestTimestamp, e)
         }
 
-        return null
+        return
```
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/VirtualNodeUpgradeOperationHandler.kt]:
```suggestion
```

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/VirtualNodeAsyncOperationHandler.kt [hunk 20]
```diff
@@ -14,5 +13,5 @@ interface VirtualNodeAsyncOperationHandler<REQUEST> {
      * @param requestId the identifier of the request, used for traceability and logging purposes
      * @param request the request itself, each implemention should handle one type
      */
-    fun handle(requestTimestamp: Instant, requestId: String, request: REQUEST): Record<*, *>?
+    fun handle(requestTimestamp: Instant, requestId: String, request: REQUEST)
```
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/VirtualNodeAsyncOperationHandler.kt:16]: We should probably amed the KDoc on line 10, as we no longer returnign the record.

**Review by vkolomeyko — COMMENTED** [review]
Overall looks fine -few minor comments.

File: testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterUtils.kt [hunk 21]
```diff
@@ -90,33 +90,31 @@ fun getOrCreateVirtualNodeFor(
...
+            val holdingId = createVNodeRequest["requestId"].textValue()
 
             // Wait for the vNode creation to propagate through the system before moving on
             eventually {
-                assertThat(
-                    vNodeList().toJson()["virtualNodes"].map {
-                        it["holdingIdentity"]["shortHash"].textValue()
-                    }.contains(holdingId)
-                )
+                val vNodeList=vNodeList().toJson()["virtualNodes"].map {
```
Comment (vkolomeyko) [inline testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterUtils.kt]: Nit:
```suggestion
                val vNodeList = vNodeList().toJson()["virtualNodes"].map {
```

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/RequestFactory.kt (hunk 16 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/RequestFactory.kt]: fixed

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/CreateVirtualNodeOperationHandler.kt (hunk 18 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/CreateVirtualNodeOperationHandler.kt]: fixed

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/VirtualNodeAsyncOperationHandler.kt (hunk 20 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/VirtualNodeAsyncOperationHandler.kt:16]: fixed

File: testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterUtils.kt (hunk 21 above)
Comment (owenstanford) [reply testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterUtils.kt]: fixed

File: components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/impl/RequestFactoryImpl.kt (hunk 17 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/factories/impl/RequestFactoryImpl.kt]: fixed

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/VirtualNodeUpgradeOperationHandler.kt (hunk 19 above)
Comment (owenstanford) [reply components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/VirtualNodeUpgradeOperationHandler.kt]: fixed

**Review by vkolomeyko — DISMISSED** [review]
LGTM

**Review by vkolomeyko — APPROVED** [review]
LGTM

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/CreateVirtualNodeOperationHandler.kt [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+        val cpiMetadata = execLog.measureExecTime("get CPI metadata") {
+            createVirtualNodeService.getCpiMetaData(request.cpiFileChecksum)
+        }
+
+        execLog.measureExecTime("check holding identity uniqueness") {
+            createVirtualNodeService.ensureHoldingIdentityIsUnique(request)
+        }
+
+        val vNodeDbs = execLog.measureExecTime("get virtual node databases") {
+            virtualNodeDbFactory.createVNodeDbs(holdingId.shortHash,request)
```
Comment (sprintengine) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/asyncoperation/handlers/CreateVirtualNodeOperationHandler.kt:53]: I think at some point we'll want to write a row in `virtual_node_operation` to record completion of each of these steps and indicate when the liquibase migrations were initiated, and integrate this with the [heartbeat mechanism](https://github.com/corda/platform-eng-design/blob/master/core/corda-5/corda-5.0/virtual-node/cpi-upgrade/cpi-upgrade-endpoint-and-logic.md#upgrade-heartbeat-mechanism). 

Atm, if the vnode creation fails, there's no way to know why (other than looking at the logs), you'll just get a 404 on the GET virtualnode endpoint. It would be useful if users can find out why their creation failed by taking the vnodeId returned by the endpoint and use it in the `/api/v1/virtualnode/status/{requestId}` to find out why an operation failed.

**Review by ronanbrowne — APPROVED** [review]
lgtm - build

**Review by conal-smith — APPROVED** [review]
LGTM - I deployed and tested a few things out locally
